File: packages/core/tasks/start.ts
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { join, resolve } from "node:path";
import { to } from "@repo/shared/lib/utils/try.ts";
// import { eq, sql } from 'drizzle-orm';
import { getStages } from "@repo/core/stages/utils/stages";
// import { taskStages, tasks } from './../../feat/tasks/table.ts';
import { readInputArgs } from "@repo/core/input/input";
import { STAGE_HANDLERS } from "../stages/index";
import {
  TaskCtx,
  readCtx,
  readPipeline,
  readTask,
  setCtx,
  setStage,
  setTask,
  writeCtx,
  writeStages,
  listStage,
  _writeCtx,
} from "@repo/core/context/context.ts";
import {
  emitLog,
  setLogContext,
  setCurrentStage,
  getStageStatuses,
  nowISO,
  // updateStageDB,
  // updateTaskDB,
} from "@repo/core/stages/utils/utils.ts";
import { log } from "@repo/util/log";
import { snapshotInput } from "./util";

export async function runPipeline(ctx: TaskCtx) {
  const taskId = ctx.task.id;
  const taskDir = ctx.task.task_dir;
  setLogContext(taskDir);
  let task = readTask(taskDir);
  mkdirSync(taskDir, { recursive: true });

  const pipeline = readPipeline(taskDir);
  const stages = getStages(pipeline);
  const targetStage = ctx.input?.targetStage;
  if (targetStage && !stages.find((s) => s === targetStage)) {
    log(`[WARN] targetStage "${targetStage}" 不在 ${pipeline} pipeline 中，忽略`);
  }

  snapshotInput(taskDir);

  setTask(taskDir, { status: "running", started_at: nowISO() });

  for (const stage of stages) {
    const handler = STAGE_HANDLERS[stage];
    if (!handler) {
      log(`[WARN] [Pipeline] No handler for stage ${stage}, skipping`);
      continue;
    }

    setStage(taskDir, stage, {
      status: "running",
      started_at: nowISO(),
      last_message: `Starting ${stage}...`,
    });
    setTask(taskDir, { status: "running", current_stage: stage, started_at: nowISO() });
    setCurrentStage(stage);
    try {
      await handler(taskDir);
      if (targetStage && stage === targetStage) {
        log(`[Pipeline] 达到目标步骤 "${targetStage}"，停止`);
        break;
      }
    } catch (err: any) {
      const msg = err.message ?? String(err);
      log(`[ERROR] [Pipeline] Stage ${stage} failed: ${msg}`);
      setStage(taskDir, stage, {
        status: "failed",
        error_message: msg,
        completed_at: nowISO(),
      });
      setTask(taskDir, { status: "failed", error_message: msg });
      throw err;
    }

    const next = readTask(taskDir);
    if (next) {
      task = next;
    }
  }

  setTask(taskDir, {
    status: "success",
    completed_at: nowISO(),
    current_stage: null,
  });
  log(`[Pipeline] Task ${taskId} completed`);
}

File: packages/core/src/cmd/tasks/start.rs
//! start 派发器 (镜像 TS `packages/core/cmd/tasks/startTask.ts` 的 `cmdStartTask`)。
//!
//! 语义: 无论是否带 `taskDir`, 都按 `input.task.url` 重新 `import_video` 生成/复用
//! 任务目录, 再从头跑完整 pipeline (`run_pipeline`)。`taskDir` 在此路径被忽略
//! (对齐 TS: start 分支忽略 taskDir, 不会续跑已有任务)。
//!
//! 与 [`crate::tasks::continue_pipeline`] 的区别: start 会 import, continue 不 import。

use crate::input::Input;
use anyhow::Context;

/// 启动新任务: 导入视频 → 跑完整 pipeline (镜像 TS `cmdStartTask`)。
pub fn start_task(input: &Input) -> anyhow::Result<()> {
    let ctx = crate::tasks::import::download::import_video(input).context("import_video 失败")?;
    println!("[cli] 导入完成, task_dir = {}", ctx.task.task_dir);
    crate::tasks::pipeline::run_pipeline(&ctx.task.task_dir).context("run_pipeline 失败")?;
    println!("[cli] 完成: {}", ctx.task.task_dir);
    Ok(())
}
